ui/src/DevUtils/MockVertx/MockVertxServer.js

```diff
@@ -0,0 +1,292 @@
+  Object.entries(config).forEach(([event, handler]) => {
+    ws.on(event, (evt) =>
+      handler({
+        data: JSON.parse(evt),
+        url,
+        ws,
+      })
+    );
+  });
+};
+
+// generator for a simple logger
+const generateLogger = (enabled = false) =>
+  enabled
+    ? (source = 'UNKNOWN_SOURCE', msg = 'Empty message logged', other) => {
+        // eslint-disable-next-line no-console
+        console.log(`${new Date()} - ${source} - ${msg}`);
+        // eslint-disable-next-line no-console
+        other && console.dir(other);
+      }
+    : () => undefined;
+
+const generateModelForConfig = ({
+  produceRate = TICK_RATE,
+  topic = TOPIC_NAME,
+  numberOfPartitions = NUMBER_OF_PARTITIONS,
+  defaultPayload = PLACEHOLDER_PAYLOAD,
+  maxOffsets = MAX_OFFSETS,
+  enableLog = false,
+  errorChance = ERROR_CHANCE,
+}) => {
+  // create in this scope a set of variables which will simulate the state
+  // of the producer/consumer
+  let consumerConnected = false;
+  let producedRecords = [];
+  let lastOffsetForPartition = new Map();
+  let mockProducerInterval;
+  let mockConsume = () => false;
+  const logger = generateLogger(enableLog);
+
+  const mockProduce = (custom = defaultPayload, ws) => {
+    const partitionBeingProducedTo = Math.floor(
+      Math.random() * numberOfPartitions
+    );
+    const offset = lastOffsetForPartition.get(partitionBeingProducedTo) || 0;
+    const produceSuccess = Math.floor(Math.random() * 100) > errorChance;
+    let response = {};
+
+    if (produceSuccess) {
+      response = generateMockSuccessShape(
+        topic,
+        partitionBeingProducedTo,
+        offset,
+        custom
+      );
+      logger('Mock Produce event', 'Producing the following record', response);
+      producedRecords.push(response);
+      lastOffsetForPartition.set(partitionBeingProducedTo, offset + 1);
+    } else {
+      logger('Mock Produce event', `Simulating produce error response`);
+      response = generateMockErrorResponseShape();
+      producedRecords.push(response);
+    }
+
+    // check if we are over our limit for offsets (simulate retention)
+    const offsetLimit = producedRecords.length - maxOffsets;
+    if (offsetLimit > 0) {
+      logger(
+        'Mock Produce event',
+        `Max offset limit (${maxOffsets}) exceeded (current size: ${producedRecords.length}, over limit by: ${offsetLimit}). Removing oldest records.`
+      );
+      producedRecords = producedRecords.slice(
+        offsetLimit,
+        producedRecords.length
+      );
+    }
+    ws.send(JSON.stringify(response)); // send a response back having 'produced'
+    mockConsume();
+  };
+
+  const handleProduceMessage = (
+    { action, custom },
+    ws,
+    intervalTimer = produceRate
+  ) => {
+    // if action start, kick off an interval to add to the to be consumed array
+    // if stop, cancel that timer
+    if (action === 'start' && !mockProducerInterval) {
```

Comment:
Clarification question - double checking new sample app sends one message on websocket to "start producing with payload x"?

Reply:
Client sends a payload ala: `{action: 'start', payload: 'something'}`. If no payload, the backend sends `Hello world`